#983 - Infusionsoft Actions & Sources

GTFalcao opened the pull request:
# New Actions
- **Create Invoice** (create payment)
- **Create Order Item** (add product to invoice)
- **Get Company**
- **Get Contact**

# New Sources
- ~~**New Appointment**~~ (removed for the time being due to a bug on Keap's side)
- **New Invoice**
- **New Order**
- **New Payment**

Closes #983

<a href="https://gitpod.io/#https://github.com/PipedreamHQ/pipedream/pull/3870"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

dylburger commented:
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible
- [ ] [Create subscription webhooks sources (vs polling sources) whenever possible](https://pipedream.com/docs/components/guidelines/#webhook-vs-polling-sources)
- [ ] [Always emit a `summary` with events](https://pipedream.com/docs/components/guidelines/#emit-a-summary)
- [ ] [Use a built-in deduping strategy wherever possible](https://pipedream.com/docs/components/guidelines/#deduping)
- [ ] When building webhook sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#webhook-sources)
- [ ] When building polling sources, [follow these best practices](https://pipedream.com/docs/components/guidelines/#polling-sources)

jcortes commented on components/infusionsoft/sources/common.ts:
```diff
@@ -0,0 +1,93 @@
+
+    let shouldTriggerEvent = true;
+
+    // If this is a hook verification request:
+    // Do not trigger an event (respond with the secret received)
+    if (hookSecret && data.method === "POST") {
+      shouldTriggerEvent = false;
+      httpResponse.headers[hookSecretName] = hookSecret;
+    };
+
+    this.http.respond(httpResponse);
+
+    // Actual event trigger
+    if (shouldTriggerEvent) {
+      const objectKeys = data.body.object_keys;
+      if (!(objectKeys instanceof Array)) {
+        throw new Error('Unknown data received from Infusionsoft webhook');
+      };
+
+      const promises: Promise<void>[] = objectKeys.map(
+        async (obj: webhookObject) =>
+          new Promise(async (resolve) => {
+            const { apiUrl, id, timestamp } = obj;
+
+            const response = await this.infusionsoft.hookResponseRequest(apiUrl);
+            const data = response.noUrl ? obj : response;
+            const summary = this.getSummary(data);
+
+            this.$emit(data, {
+              id,
+              summary,
+              ts: new Date(timestamp).valueOf(),
+            });
+
+            resolve();
+          })
+      );
+
+      await Promise.allSettled(promises);
+    }
```
How about this refactor?
```suggestion
    // If this is a hook verification request:
    // Do not trigger an event (respond with the secret received)
    if (hookSecret && data.method === "POST") {
      return this.http.respond({
        headers: {
          hookSecretName: hookSecret,
        },
        status: 200,
      });
    }

    const { object_keys: objectKeys } = data.body;
    if (!Array.isArray(objectKeys)) {
      throw new Error("Unknown data received from Infusionsoft webhook");
    }

    const promises: Promise<{
      obj: webhookObject;
      response: any;
    }>[] =
      objectKeys.map(async (obj: webhookObject) => ({
        obj,
        response: await this.infusionsoft.hookResponseRequest(obj.apiUrl),
      }));

    const result = await Promise.all(promises);

    result.forEach(({
      obj, response,
    }) => {
      const data = response.noUrl
        ? obj
        : response;
      const summary = this.getSummary(data);
      this.$emit(data, {
        id: obj.id,
        summary,
        ts: new Date(obj.timestamp).valueOf(),
      });
    });
```

GTFalcao replied:
Great suggestions! The promise wrapping was definitely not ideal and this is much better.

I've `return`'ed when aborting the event emitting, instead of using a variable, as you suggested, however I kept `this.http.respond` as well after that - as the workflow needs to respond with 200 both when it is validating the webhook, and when receiving an event.

jcortes reviewed (commented):
Hi @GTFalcao great work. I've just added a few request changes related to eslint errors and some refactoring code. Other than that it's looking great. Thanks

GTFalcao commented on components/infusionsoft/sources/common.ts:
```diff
@@ -0,0 +1,93 @@
+import infusionsoft from "../app/infusionsoft.app";
+import { SourceHttpRunOptions } from "@pipedream/types";
```
That's likely because it was a recent addition to the types package, on `v0.1.4` - previously on `v0.1.3` there was only `SourceRunOptions`. 

https://github.com/PipedreamHQ/pipedream/blob/9a72c1c141865b4126e3e0096452e5c068666ee4/types/src/index.ts#L239

I don't get this error, SourceHttpRunOptions is correctly typed for me - you might need to `npm run build` to build that TS package update

GTFalcao commented on components/infusionsoft/actions/create-order-item/create-order-item.ts:
```diff
@@ -0,0 +1,55 @@
+  props: {
+    infusionsoft,
+    orderId: {
+      propDefinition: [infusionsoft, "orderId"],
```
I was used to all indentation patterns being applied by husky on pre-commit, so I wasn't checking on those - but looks like that's not being done for .ts files just yet!

I've applied eslint --fix to correct all these issues. Thanks for the heads up and suggestion!

GTFalcao commented on components/infusionsoft/app/infusionsoft.app.ts:
```diff
@@ -0,0 +1,214 @@
+        \\
+        Alternatively, you can provide a custom *Contact ID*.`,
+      async options() {
+        const contacts: contact[] = await this.listContacts();
```
Definitely! Thanks for the reminder!

I've applied PascalCase to all the defined types

GTFalcao commented on components/infusionsoft/app/infusionsoft.app.ts:
```diff
@@ -0,0 +1,214 @@
+      });
+
+      return response.data;
+    },
```
Thanks for the suggestion - applied this

GTFalcao commented on components/infusionsoft/app/infusionsoft.app.ts:
```diff
@@ -0,0 +1,214 @@
+import { defineApp } from "@pipedream/types";
+import axios from "axios";
```
Good call - I used the RSS components as a base and did not remember to switch to `platform` axios.

Applied this, and passed `$` from all the actions down to their http request methods.

GTFalcao commented:
Thanks for your very helpful review @jcortes ! I've addressed all the notes you left and moved this back to the review queue

jcortes reviewed (commented):
@GTFalcao Great work! Thanks for making all theses changes. Ready for QA!

DilanAthukorala commented:
@GTFalcao Following Issues / improvements are observed

1. for create payment action, payment amount and payment method props should be mandatory but currently it has been defined as optional. 
![image](https://user-images.githubusercontent.com/102980146/182040801-0ea48db6-34a6-4bcd-83fc-17aa462aa7ad.png)


2. No events are emitted for New Appointment source after creating appointments
![image](https://user-images.githubusercontent.com/102980146/182041403-9e59e8d7-58f4-4753-bada-b2adddae3e9d.png)

GTFalcao commented:
I've changed the payment method and amount props to required. They were initially set to optional as the API does not label them as required, but I can see why they should be.

As for the 'New Appointment' source, it seems Keap's app/interface is not triggering the event correctly, as when I created the appointment via their API, it triggered as expected. Attaching screenshots below.

I reported this bug to Keap and, for the time being, removed the 'New Appointment' source from this pull request, so that the other components can be released instead of being blocked by this issue.

![image](https://user-images.githubusercontent.com/48412907/182248304-96718d76-150a-4e53-ac40-6bf16f17d320.png)
![image](https://user-images.githubusercontent.com/48412907/182248333-1a496d94-932e-4a4c-993f-0c948ca6f9ff.png)

DilanAthukorala commented:
Verification completed and this is ready for release.
